CorsixTH/Lua/dialogs/place_staff.lua

```diff
@@ -77,15 +77,26 @@ end
 
 local flag_cache = {}
 local flag_altpal = 16
+-- Is this coordinate a valid tile for this staff member?
+local function isValidPlace(self)
+  self.world.map.th:getCellFlags(self.tile_x, self.tile_y, flag_cache)
+  local room = self.world:getRoom(self.tile_x, self.tile_y)
+  -- Is the tile inside the player's hospital building?
+  local isplayer = self.ui.hospital:getPlayerIndex() == flag_cache.owner and flag_cache.hospital
+  -- On a tile humanoids can walk on?
+  local walkable = flag_cache.passable and (not room and true or not room.crashed)
+  -- and in an area the regular staff (doctors, nurses and handymen) can go?
+  local staffable = (self.allow_in_rooms or flag_cache.roomId == 0)
+  local valid = walkable and staffable
+  -- Or is it a receptionist placed on the centre tile of an unstaffed desk?
+  local reception = self.staff and self.staff.profile.humanoid_class == "Receptionist" and
+      self.world:getObject(self.tile_x, self.tile_y, "reception_desk")
+  return isplayer and (reception or valid)
+end
```

Comment:
Breaking up that long condition is great!

Try adding a few empty lines in there, so you get 'blocks' of code that are closely connected. It makes reading it much simpler. [EDIT: Below I suggest to rewrite the whole thing, consider this afterwards.]
Also:
- Line 80 is not complete, it should have luadoc explanations
- `isPlayer` seems the wrong name for an inside-tile check.
- Move 85 closer to83, so 'room things' are near each other.
- If you rename `walkable` to `walkable_tile`, line 86 becomes redundant. This may be possible to other lines too.
- `allow_in_rooms` makes no sense as name at all, but that's not part of this patch, probably.
- `valid` is too wide, it doesn't make clear what validness you decided here.

The final `return` statement possibly invalidates several computations, ie, you're wasting CPU time on doing computations that you don't need.

Finally, line 92 is a whole different branch of computations, not related at all to any previous computed value, it's a big twist in the plot :) 

In all, the direction is good I think, it just needs a second sweep. I'd suggest something along the lines of
``` lua
local function isValidPlace(self)
  -- Insert code to compute isPlayer here
  if not isPlayer return false end

  -- Insert code for reception here
  if reception then return true end

  -- Insert code for valid here
  return valid
end
```